### PR TITLE
MAINT: add npy_gil_error to acquire the GIL and set an error

### PR DESCRIPTION
**Pull request by ngoldbaum**
This adds `npy_gil_error`, a helper I wrote for stringdtype/NEP 55. I'm making a PR here separately from the stringdtype PR to ease review.

This function acquires the GIL, sets an error using `PyErr_FormatV`, and releases the GIL. I've replaced all the places in NumPy that follow this pattern, ignoring a few places that do something more complicated than just setting an error.

I'm not using the fact that `npy_gil_error` is a varargs function in numpy yet, but do want to use that for stringdtype. It works locally both without any varargs and with varargs.

This is purely a refactoring and should have no user-visible impacts.

### CODE REVIEW

File: numpy/_core/src/common/gil_utils.c [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+NPY_NO_EXPORT void
+npy_gil_error(PyObject *type, const char *format, ...)
+{
+    va_list va;
+    va_start(va, format);
+    NPY_ALLOW_C_API_DEF;
+    NPY_ALLOW_C_API;
+    if (!PyErr_Occurred()) {
+        PyErr_FormatV(type, format, va);
```
Comment (charris) [inline numpy/_core/src/common/gil_utils.c:19]: Note that PyPy 3.9 does not support PeErr_FormatV.

**Comment by ngoldbaum** [comment]
Looks like [nanobind has a workaround for this](https://github.com/wjakob/nanobind/blob/eb308de0e3099e5abd8e2e9b60a8e0af81ee93db/src/error.cpp#L273C1-L273C1).

File: numpy/_core/src/common/gil_utils.c [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+#include <numpy/ndarraytypes.h>
+
+#include <stdarg.h>
+
+NPY_NO_EXPORT void
+npy_gil_error(PyObject *type, const char *format, ...)
+{
+    va_list va;
+    va_start(va, format);
```
Comment (seberg) [inline numpy/_core/src/common/gil_utils.c:15]: This looks good to me, thanks.

One thing though: I think at least technically every `va_start` call must be followed by a `va_end(va)` call before returning.

Comment (ngoldbaum) [reply numpy/_core/src/common/gil_utils.c:15]: Thanks!

**Comment by seberg** [comment]
I quite like this as a helper and it's useful for the string PR, so lets put this in, thanks.

**Pull request merged**